tidy3d/components/medium.py

```diff
@@ -1546,6 +1553,146 @@ def to_medium(self) -> Medium:
+        """
+        return f_rad / 2 / np.pi
+
+    @staticmethod
+    def _Hz_to_angular_freq(f_hz: float):
+        """Convert frequency in unit of Hz to rad/s.
+
+        Parameters
+        ----------
+        f_hz : float
+            Frequency in unit of Hz
+        """
+        return f_hz * 2 * np.pi
+
+    @cached_property
+    def _imag_ep_extrema(self) -> Tuple[ArrayFloat1D]:
+        """Extrema (in unit of rad/s) of Im[eps]."""
+
+        def _extrema_loss_freq_finder(areal, aimag, creal, cimag):
+            """For each pole, find frequencies for the extrema of Im[eps]"""
+
+            a_square = areal**2 + aimag**2
+            alpha = creal
+            beta = creal * (areal**2 - aimag**2) + 2 * cimag * areal * aimag
+            mus = 2 * (areal**2 - aimag**2)
+            nus = a_square**2
+
+            numerator = np.array([0])
+            denominator = np.array([1])
+            for i in range(len(creal)):
+                numerator_i = np.array(
+                    [
+                        -alpha[i],
+                        alpha[i] * mus[i] - 3 * beta[i],
+                        3 * alpha[i] * nus[i] - beta[i] * mus[i],
+                        beta[i] * nus[i],
+                    ]
+                )
+                denominator_i = np.array(
+                    [1, 2 * mus[i], 2 * nus[i] + mus[i] ** 2, 2 * mus[i] * nus[i], nus[i] ** 2]
+                )
+                # to avoid divergence, let's renormalize
+                if np.abs(alpha[i]) > 1:
+                    numerator_i /= alpha[i]
+                    denominator_i /= alpha[i]
+
+                # n/d + ni/di = (n*di+d*ni)/(d*di)
+                n_di = np.polymul(numerator, denominator_i)
+                d_ni = np.polymul(denominator, numerator_i)
+                numerator = np.polyadd(n_di, d_ni)
+                denominator = np.polymul(denominator, denominator_i)
+
+            roots = np.sqrt(np.roots(numerator) + 0j)
+            # cutoff to determine if it's a real number
+            r_real = roots.real[np.abs(roots.imag) / (np.abs(roots) + fp_eps) < fp_eps]
+            return r_real[r_real > 0]
+
+        try:
+            # convert to the unit eV for better numerical handling, since those quantities
+            # will be ~ 1 in photonics
+            poles, residues = zip(*self._angular_freq_to_eV(np.array(self.poles)))
+            poles = np.array(poles)
+            residues = np.array(residues)
+            extrema_freq = _extrema_loss_freq_finder(
+                poles.real, poles.imag, residues.real, residues.imag
+            )
+            return PoleResidue._eV_to_angular_freq(extrema_freq)
+        except np.linalg.LinAlgError:
+            log.warning(
+                "'LinAlgError' in computing Im[eps] extrema. Im[eps] upper bound not accurate."
+            )
+            return np.array([])
+
+    def _imag_ep_extrema_with_samples(self, all_frequency_range: bool = False) -> ArrayFloat1D:
```

Comment:
I don't fully understand this function. It looks like it only keeps extrema within the frequency range, and also returns the frequency range in addition to those extrema. Previously, I was checking min_loss and max_loss at all extrema, regardless of whether they were in this frequency range. Then, I was also checking over the frequency range in case the extrema finder failed. While I can see only wanting to enforce max_loss over a certain frequency range, I thought we wanted passivity at all frequencies.

Reply:
Now fit_fast calls the extrema finder function directly. This function is only used to find max(min)_loss in the frequency_range.

Reply:
thanks!